rfcs/20190829-tfx-container-component-execution.md

```diff
@@ -0,0 +1,555 @@
+    type: object
+    properties:
+      code:
+        type: string
+        enum: [PERMANENT_ERROR, RETRYABLE_ERROR]
+      message:
+        type: string
+```
+
+The output.json file is optional, but if user’s container writes to the file. It
+overrides the default handling of k8s pod launcher. Here are the explanation of
+the output fields:
+
+*   error_status: it tells the executor whether it should retry or fail.
+*   outputs and exec_properties: they will be used to override the execution and
+    output artifact metadata in MLMD.
+
+The output interfaces relies on `BaseComponentLauncher` to update states back to
+MLMD from executor.
+
+### Auth context resolution
+
+K8s pod launcher internally uses k8s python client. The auth context resolution
+logic is as follows:
+
+1.  If the current env is in a cluster, uses `load_incluster_config` to load k8s
+    context.
+1.  If not, use default k8s active context to connect to remote cluster.
+
+### Pod launcher resiliency
+
+In this design section, we focused more on the launcher resiliency under
+`KubeflowDAGRunner`. In `AirflowDAGRunner`, the launcher code is running in the
+same process of Airflow orchestrator which we rely on Airflow to ensure its
+resiliency. `BeamDAGRunner`, however, is considered mainly for local testing
+purpose and we won't add support for it to be resiliency.
+
+In `KubeflowDAGRunner`, a pipeline step will create two pods in order to execute
+user’s container:
+
+*   A launcher pod which contains driver, k8s pod launcher and publisher code.
+*   A user pod with user’s container.
+
+Pod in k8s is not resilient by itself. We will use Argo’s retry feature to make
+launcher pod to be partially resilient. The details are as follows:
+
+*   Each argo launcher step will be configured with a default retry count.
+*   Argo will retry the step in case of failure no matter what type of error.
+*   The launcher container will create a tmp workdir from `pipeline_root`.
+*   It will keep intermediate results like created pod ID in the tmp workdir.
+*   The k8s pod launcher will be implemented in a way that it will resume the
+    operation based on the intermediate results in the tmp workdir.
+*   The launcher will also record a permanent failure data in the tmp workdir so
+    it won’t resume the operation in case of non-retriable failures.
+
+### Default retry strategy
+
+K8s pod launcher supports exponential backoff retry. This strategy applies to
+all runners which can support k8s pod launcher. Docker launcher are not in the
+scope of the design as it is mainly for local development use case.
+
+The retry only happens if the error is retriable. An error is retriable only
+when:
+
+*   It’s a transient error code from k8s pod API.
+*   The output.json file from artifact store indicates it’s a retriable error.
+*   The pod get deleted (For example: GKE preemptible pod feature).
+
+### Log streaming
+
+Container launcher streams log from user’s docker container or k8s pod through
```

Comment:
```suggestion
The container launcher streams the log from user’s docker container or k8s pod through the
```

Reply:
_user’s Docker container_  - caps for consistency